book/src/specification/logical.md

```diff
@@ -1,3 +1,788 @@
+> using don't-cares for the variants not selected through the tag field.
+> However, there are a number of downsides to such a sparse union approach.
+>
+>  - For unions with many variants, the `data` signal of the corresponding
+>    stream becomes very wide.
+>  - Any streams nested inside the union variants would need to be stuffed
+>    with don't-care *transfers* as well, which may result in additional
+>    control logic, delay, area, etc..
+>
+> While such sparse unions have their place, dense unions can be more suitable
+> for a given application. These could also be represented without a
+> \\(\textrm{Union}\\) node, but not in a nicely recursive manner without loss
+> of information about the variant types. Therefore, (dense) unions were made a
+> first-class type through \\(\textrm{Union}\\).
+>
+> \\(\textrm{Union}\\)s can also be used to elegantly describe nullable types.
+> The pattern for that is simply \\(\textrm{Union}(\textrm{Null}, T)\\). In
+> this case, the union's tag field acts like a validity bit.
+
+The names cannot contain two or more consecutive underscores.
+
+> Double underscores are used by Tydi as a hierarchy separator when flattening
+> structures for the canonical representation. If not for the above rule,
+> something of the form
+> \\(\textrm{Union}(\textrm{'a'}:\textrm{Union}(\textrm{'b__c'}:\textrm{Stream}...),\textrm{'a__b'}:\textrm{Union}(\textrm{'c'}:\textrm{Stream}...))\\)
+> would result in a name conflict: both physical streams would be named `a__b__c`.
+
+The names cannot start or end with an underscore.
+
+The names cannot start with a digit.
+
+The names cannot be empty.
+
+The names must be case-insensitively unique within the group.
+
+> The above requirements on the name mirror the requirements on the field names
+> for the physical streams.
+
+### Operations on logical stream
+
+This section defines some operations that may be performed on logical stream
+types, most importantly the \\(\textrm{synthesize}()\\) function, which returns
+the signals and physical streams for a given type.
+
+#### Null detection function
+
+This section defines the function
+\\(\textrm{isNull}(T_{in}) \rightarrow \textrm{Bool}\\), which returns
+true if and only if \\(T_{in}\\) does not result in any signals.
+
+The function returns true if and only if one or more of the following
+rules match:
+
+ - \\(T_{in} = \textrm{Stream}(T_e, n, d, s, c, r, T_u, x)\\),
+   \\(\textrm{isNull}(T_e)\\) is true, \\(\textrm{isNull}(T_u)\\) is true,
+   and \\(x\\) is false;
+
+ - \\(T_{in} = \textrm{Null}\\);
+
+ - \\(T_{source} = \textrm{Group}(N_1: T_1, N_2: T_2, ..., N_n: T_n)\\) and
+   \\(\textrm{isNull}(T_i)\\) is true for all \\(i \in (1, 2, ..., n)\\); or
+
+ - \\(T_{source} = \textrm{Union}(N: T)\\) and \\(\textrm{isNull}(T)\\) is true.
+
+#### Split function
+
+This section defines the function
+\\(\textrm{split}(T_{in}) \rightarrow \textrm{SplitStreams}(T_{signals}, N_1 : T_1, N_2 : T_2, ..., N_n : T_n)\\),
```

Comment:
I feel like a definition of what SplitStreams is, is missing

Reply:
b5b9bdb rewords this to hopefully be more clear, and also adds a section describing the mathematical notation in general.